pkg/runtime/interfaces.go

```diff
@@ -139,29 +148,41 @@ type StorageSerializer interface {
 
 	// EncoderForVersion returns an encoder that ensures objects being written to the provided
 	// serializer are in the provided group version.
-	// TODO: take multiple group versions
-	EncoderForVersion(serializer Encoder, gv unversioned.GroupVersion) Encoder
+	EncoderForVersion(serializer Encoder, gv GroupVersioner) Encoder
 	// DecoderForVersion returns a decoder that ensures objects being read by the provided
 	// serializer are in the provided group version by default.
-	// TODO: take multiple group versions
-	DecoderToVersion(serializer Decoder, gv unversioned.GroupVersion) Decoder
+	DecoderToVersion(serializer Decoder, gv GroupVersioner) Decoder
+}
+
+// NestedObjectEncoder is an optional interface that objects may implement to be given
```

Comment:
Because you don't always want the same convertor to be used.  For example, the majority of runtime.Unknown -> runtime.RawExtension conversions could be generic to context (just does a straight up decode).  But specific cases actually need to handle objects differently - for instance Templates don't actually want to decode objects because type marshaling would break the template replacement.  Templates want to use the unstructured decoder, which preserves the exact syntax of the owner.

So basically, each top level object can define a distinct behavior.